File: projectingAndRotating.py
import math as m

START_POSITION = [0, 0]

def DegreeToRadian(D):
    return D * m.pi / 180


def RadianToDegree(R):
    return 180 * R / m.pi


def project(x, y, z, F):
    if F <= z:
        print("Error! Can't project a point behind the viewing point.\n")
        return
    else:
        px = F * x / (F - z)
        py = F * y / (F - z)
        P = [px, py]

    return P


def Rotate2D(x, y, angle):
    # Define the 2D rotation matrix
    R = [
        [m.cos(angle), -m.sin(angle)],
        [m.sin(angle), m.cos(angle)]
    ]

    # Apply the rotation matrix to the coordinates
    rx = R[0][0] * x + R[0][1] * y
    ry = R[1][0] * x + R[1][1] * y

    return [round(rx, 8), round(ry, 8)]

def Rotate3DInX(x, y, z, A):
    R = [
        [1, 0, 0],
        [0, m.cos(A), -m.sin(A)],
        [0, m.sin(A), m.cos(A)]
    ]
    
    rx = x * R[0][0] + y * R[0][1] + z * R[0][2]
    ry = x * R[1][0] + y * R[1][1] + z * R[1][2]
    rz = x * R[2][0] + y * R[2][1] + z * R[2][2]
    return [round(rx, 8), round(ry, 8), round(rz, 8)]

def Rotate3DInY(x, y, z, A):
    R = [
        [m.cos(A), 0, m.sin(A)],
        [0, 1, 0],
        [-m.sin(A), 0, m.cos(A)]
    ]
    
    rx = x * R[0][0] + y * R[0][1] + z * R[0][2]
    ry = x * R[1][0] + y * R[1][1] + z * R[1][2]
    rz = x * R[2][0] + y * R[2][1] + z * R[2][2]
    return [round(rx, 8), round(ry, 8), round(rz, 8)]

def Rotate3DInZ(x, y, z, A):
    R = [
        [m.cos(A), -m.sin(A), 0],
        [m.sin(A), m.cos(A),  0],
        [0, 0, 1]
    ]
    
    rx = x * R[0][0] + y * R[0][1] + z * R[0][2]
    ry = x * R[1][0] + y * R[1][1] + z * R[1][2]
    rz = x * R[2][0] + y * R[2][1] + z * R[2][2]
    return [round(rx, 8), round(ry, 8), round(rz, 8)]

File: projectingAndRotating.c
#include<stdio.h>
#define _USE_MATH_DEFINES
#include<math.h>

typedef struct
{
    double x;
    double y;
} Coords2d;

typedef struct
{
    double x;
    double y;
    double z;
} Coords3d;

double DegreeToRadian(double D) {
    return D * M_PI / 180;
}

double RadianToDegree(double R) {
    return 180 * R / M_PI;
}

Coords2d project(double x, double y, double z, double F) {
    if (F <= z) {
        printf("Error! Not possible project a point that is behind the viewing point!\n");
    }
    else {
        Coords2d p;
        p.x = x * F / (F - z);
        p.y = y * F / (F - z);
        return p;
    }
}

Coords2d Rotate2D(double x, double y, double A) {
    Coords2d r;
    r.x = x * cos(A) - y * sin(A);
    r.y = x * sin(A) + y * cos(A);
    return r;
}

Coords3d Rotate3DInX(double x, double y, double z, double A) {
    double R[3][3] = {
        {1, 0, 0},
        {0, cos(A), -sin(A)},
        {0, sin(A), cos(A)}
    };
    Coords3d r;
    r.x = R[0][0] * x + R[0][1] * y + R[0][2] * z;
    r.y = R[1][0] * x + R[1][1] * y + R[1][2] * z;
    r.z = R[2][0] * x + R[2][1] * y + R[2][2] * z;
    return r;
}

Coords3d Rotate3DInY(double x, double y, double z, double A) {
    double R[3][3] = {
        {cos(A), 0, sin(A)},
        {0, 1, 0},
        {-sin(A), 0, cos(A)}
    };
    Coords3d r;
    r.x = R[0][0] * x + R[0][1] * y + R[0][2] * z;
    r.y = R[1][0] * x + R[1][1] * y + R[1][2] * z;
    r.z = R[2][0] * x + R[2][1] * y + R[2][2] * z;
    return r;
}

Coords3d Rotate3DInZ(double x, double y, double z, double A) {
    double R[3][3] = {
        {cos(A), -sin(A), 0},
        {sin(A), cos(A), 0},
        {0, 0, 1}
    };
    Coords3d r;
    r.x = R[0][0] * x + R[0][1] * y + R[0][2] * z;
    r.y = R[1][0] * x + R[1][1] * y + R[1][2] * z;
    r.z = R[2][0] * x + R[2][1] * y + R[2][2] * z;
    return r;
}

int main () {
    double r = DegreeToRadian(45);
    Coords3d R = Rotate3DInX(20, 0, 20, r);
    printf("%lf  %lf  %lf\n", R.x, R.y, R.z);
    return 0;
}
